### PR TITLE
Add Semaphore.Resource, for creating module semaphore's

### PR DESCRIPTION
**Pull request by tim-smart**
Better as a seperate library?

Can be used to limit access to expensive resources.

### CODE REVIEW

**Review by ihumanable — COMMENTED** [review]
I need to think some about this, since the semaphore operations have to switch into the Semaphore.Resource GenServer server context this is a bottleneck and orders of magnitude slower than using the semaphore directly.

As a library maintainer there are 2 questions in my head

1. Is there a better way to get the desired effect?
2. Does adding this affordance potentially lead people to adopting patterns that will not scale well and then they get upset.

File: lib/semaphore/resource.ex [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  def start_link(name, max, opts \\ []) when is_atom(name) do
+    GenServer.start_link(__MODULE__, {name, max}, opts)
+  end
+
+  @doc """
+  Acquires the underlying semaphore. If it is unavailable, it will block until
+  the semaphore can be acquired.
+  """
+  @spec acquire(pid(), timeout()) :: :ok
+  def acquire(pid, timeout \\ :infinity) do
```
[the remark this answers is not in the document]
Comment (tim-smart) [reply lib/semaphore/resource.ex:57]: Ah this is great. I had no idea the `:queue` module existed hah. Will definitely be using that again.

Will push a refactor soon.

**Comment by tim-smart** [comment]
> I need to think some about this, since the semaphore operations have to switch into the Semaphore.Resource GenServer server context this is a bottleneck and orders of magnitude slower than using the semaphore directly.
> 
> As a library maintainer there are 2 questions in my head
> 
> 1. Is there a better way to get the desired effect?
> 2. Does adding this affordance potentially lead people to adopting patterns that will not scale well and then they get upset.

Valid questions. Let me know if you answer the first question haha!

Your second question re-affirms the seperate library thought I think. There are multiple ways you can react to a `{:error, :max}`, depending on what guarantees your application is promising.

**Comment by tim-smart** [comment]
@ihumanable Thanks for your feedback. I'll close this and move it somewhere else :)

**Pull request closed without merging**